### PR TITLE
[staging] python3: 3.8 -> 3.9

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
closes: #116304
closes: #124049

https://hydra.nixos.org/jobset/nixpkgs/python-unstable

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dotlambda** [comment]
We should add an entry to the release notes.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
At least https://github.com/NixOS/nixpkgs/blob/fcf99ed10f2bc6e8fc1cc9ef4613770038311421/doc/languages-frameworks/python.section.md#L842 needs updating but we should probably convert all python38 to python3 in that file.

**Comment by dotlambda** [comment]
@zimbatm It looks like `google-compute-engine` doesn't require `boto` in newer versions: https://github.com/GoogleCloudPlatform/compute-image-packages/blob/506b9a0dbffec5620887660cd42c57b3cbbadba6/packages/python-google-compute-engine/setup.py#L23-L26
Could you look into updating it?

File: pkgs/development/python-modules/btrees/default.nix [hunk 1]
```diff
@@ -5,27 +5,44 @@
...
+    zope_interface
+  ];
+
+  checkInputs = [
+    zope_testrunner
+  ];
+
+  checkPhase = ''
+    runHook preCheck
+    python -m zope.testrunner --test-path=src --auto-color --auto-progress
```
Comment (dotlambda) [inline pkgs/development/python-modules/btrees/default.nix:35]:
```suggestion
    ${python.interpreter} -m zope.testrunner --test-path=src --auto-color --auto-progress
```

File: pkgs/development/python-modules/btrees/default.nix [hunk 2]
```diff
@@ -5,27 +5,44 @@
...
-    substituteInPlace BTrees/tests/common.py \
-      --replace "testShortRepr" "no_testShortRepr"
-  '';
-
   src = fetchPypi {
     inherit pname version;
     sha256 = "d33323655924192c4ac998d9ee3002e787915d19c1e17a6baf47c9a63d9556e3";
   };
 
+  buildInputs = [
```
Comment (dotlambda) [inline pkgs/development/python-modules/btrees/default.nix:20]:
```suggestion
  nativeBuildInputs = [
```
?

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/btrees/default.nix (hunk 1 above)
Comment (mweinelt) [reply pkgs/development/python-modules/btrees/default.nix:35]: Oops, pushed a half-done change apparently. Thanks!

File: pkgs/development/python-modules/btrees/default.nix (hunk 2 above)
Comment (mweinelt) [reply pkgs/development/python-modules/btrees/default.nix:20]: Good question.

Comment (mweinelt) [reply pkgs/development/python-modules/btrees/default.nix]: It's actually a test dependency.

Comment (jbaum98) [reply pkgs/development/python-modules/btrees/default.nix]: I think there is `checkInputs`?

Comment (mweinelt) [reply pkgs/development/python-modules/btrees/default.nix]: Yep, updated in https://github.com/NixOS/nixpkgs/pull/128634/commits/dfd233ce45c08a9543742d69c6581d7683a6a376.

**Comment by jonringer** [comment]
merging, going to fix remaining regression on `staging-next`

**Comment by nh2** [comment]
@jonringer I just read https://github.com/NixOS/nixpkgs/commit/1e47dd9cb847f57ec859c071d6c697270a53d1e7 and wonder if the numbers really support turning LTO.

Filtering only to the `Significant rows`, there appears to be more slowdown than speedup:

```
faster 1.03x | 2to3
faster 1.03x | richards
faster 1.06x | mako
faster 1.06x | python_startup_no_site
faster 1.08x | python_startup

slower 1.02x | logging_silent
slower 1.03x | pickle_pure_python
slower 1.04x | scimark_monte_carlo
slower 1.05x | pickle
slower 1.05x | scimark_lu
slower 1.08x | float
slower 1.09x | regex_effbot
slower 1.18x | regex_dna
```

**Comment by jonringer** [comment]
@nh2 possibly, I think we had to be conservative about optimizations, as some affected reproducibility

**Pull request merged**